File: Modules/module_demo/src/main/java/com/example/william/my/module/demo/presenter/ArticlePresenter.java
package com.example.william.my.module.demo.presenter;

import com.example.william.my.module.demo.bean.ArticleDetailBean;
import com.example.william.my.module.demo.contract.ArticleContract;
import com.example.william.my.module.demo.data.ArticleDataSource;
import com.example.william.my.module.demo.data.ArticleRepository;
import com.example.william.my.module.demo.utils.CheckUtils;

import java.util.List;

public class ArticlePresenter implements ArticleContract.Presenter {

    private final ArticleRepository mArticleRepository;

    private final ArticleContract.View mArticleView;

    private int mPage;

    public ArticlePresenter(ArticleRepository articleRepository, ArticleContract.View view) {
        mArticleRepository =
                CheckUtils.checkNotNull(articleRepository, "ArticleRepository cannot be null!");
        mArticleView =
                CheckUtils.checkNotNull(view, "ArticleView cannot be null!");

        this.mArticleView.setPresenter(this);
    }

    @Override
    public void queryArticleList() {
        mPage = 0;
        queryArticleList(mPage);
    }

    @Override
    public void loadArticleList() {
        mPage++;
        queryArticleList(mPage);
    }

    private void queryArticleList(int page) {
        mArticleRepository.getArticleList(page, new ArticleDataSource.LoadArticleCallback() {
            @Override
            public void onArticleLoaded(List<ArticleDetailBean> articles) {
                mArticleView.showArticles(mPage == 0, articles);
            }

            @Override
            public void onDataNotAvailable() {
                if (mPage == 0) {
                    mArticleView.showEmptyView();
                }
            }

            @Override
            public void onFailure(String msg) {
                mArticleView.showToast(msg);
            }
        });
    }


    @Override
    public void start() {

    }
}

File: modules/module_sample/src/main/kotlin/com/example/william/my/module/sample/frame/presenter/ArticlePresenter.kt
package com.example.william.my.module.sample.frame.presenter

import com.example.william.my.bean.data.ArticleDetailBean
import com.example.william.my.bean.repo.ArticleRepository
import com.example.william.my.bean.repo.ArticleDataSource.LoadArticleCallback
import com.example.william.my.module.sample.frame.contract.ArticleContract

class ArticlePresenter(
    private val mArticleRepository: ArticleRepository,
    private val mArticleView: ArticleContract.View
) : ArticleContract.Presenter {

    override fun queryArticle(page: Int) {
        mArticleRepository.getArticle(page, object : LoadArticleCallback {
            override fun showLoading() {

            }

            override fun onFailure(msg: String) {
                mArticleView.onDataFail()
                mArticleView.showToast(msg)
            }

            override fun onArticleLoaded(articles: List<ArticleDetailBean>) {
                mArticleView.showArticles(articles)
            }
        })
    }

    override fun start() {

    }

    override fun clear() {

    }
}